python/ray/data/_internal/pandas_block.py

```diff
@@ -267,28 +289,133 @@ def sum_of_squared_diffs_from_mean(
 
     def sort_and_partition(
         self, boundaries: List[T], key: "SortKeyT", descending: bool
-    ) -> List["pandas.DataFrame"]:
-        # TODO (kfstorm): A workaround to pass tests. Not efficient.
-        delegated_result = BlockAccessor.for_block(self.to_arrow()).sort_and_partition(
-            boundaries, key, descending
-        )
-        return [BlockAccessor.for_block(_).to_pandas() for _ in delegated_result]
+    ) -> List[Block[T]]:
+        if len(key) > 1:
+            raise NotImplementedError(
+                "sorting by multiple columns is not supported yet"
+            )
+
+        if self._table.shape[0] == 0:
+            # If the pyarrow table is empty we may not have schema
+            # so calling sort_indices() will raise an error.
+            return [self._empty_table() for _ in range(len(boundaries) + 1)]
+
+        col, _ = key[0]
+        table = self._table.sort_values(by=col, ascending=not descending)
+        if len(boundaries) == 0:
+            return [table]
+
+        partitions = []
+        # For each boundary value, count the number of items that are less
+        # than it. Since the block is sorted, these counts partition the items
+        # such that boundaries[i] <= x < boundaries[i + 1] for each x in
+        # partition[i]. If `descending` is true, `boundaries` would also be
+        # in descending order and we only need to count the number of items
+        # *greater than* the boundary value instead.
+        if descending:
+            num_rows = len(table[col])
+            bounds = num_rows - table[col].searchsorted(
+                boundaries, sorter=np.arange(num_rows - 1, -1, -1)
+            )
+        else:
+            bounds = table[col].searchsorted(boundaries)
+        last_idx = 0
+        for idx in bounds:
+            # Slices need to be copied to avoid including the base table
```

Comment:
Is this comment relevant for pandas table as well?

Reply:
It's not, let me remove that!